packet/checksum.go

```diff
@@ -117,26 +121,26 @@ func CalculatePseudoHdrIPv6UDPCksum(hdr *IPv6Hdr, udp *UDPHdr) uint16 {
 // checksum for required pseudo-header and writes result to correct place. This
 // is required for checksum compute by hardware offload.
 func SetHWOffloadingHdrChecksum(p *Packet) {
-	ipv4, ipv6, _ := p.ParseAllKnownL3()
+	ipv4, ipv6, _ := p.ParseAllKnownL3CheckVLAN()
 	if ipv4 != nil {
-		p.GetIPv4().HdrChecksum = 0
+		ipv4.HdrChecksum = 0
 		tcp, udp, icmp := p.ParseAllKnownL4ForIPv4()
 		if tcp != nil {
-			p.GetTCPForIPv4().Cksum = SwapBytesUint16(CalculatePseudoHdrIPv4TCPCksum(p.GetIPv4()))
+			p.GetTCPForIPv4().Cksum = SwapBytesUint16(CalculatePseudoHdrIPv4TCPCksum(ipv4))
 		} else if udp != nil {
-			p.GetUDPForIPv4().DgramCksum = SwapBytesUint16(CalculatePseudoHdrIPv4UDPCksum(p.GetIPv4(), p.GetUDPForIPv4()))
+			p.GetUDPForIPv4().DgramCksum = SwapBytesUint16(CalculatePseudoHdrIPv4UDPCksum(ipv4, udp))
 		} else if icmp != nil {
 			// calculate full software checksum for icmp, cause there is no
 			// hardware calculation for icmp packets.
-			p.GetICMPForIPv4().Cksum = SwapBytesUint16(CalculateIPv4ICMPChecksum(p.GetIPv4(), p.GetICMPForIPv4(),
-				unsafe.Pointer(uintptr(unsafe.Pointer(p.GetICMPForIPv4())) + ICMPLen)))
+			p.GetICMPForIPv4().Cksum = SwapBytesUint16(CalculateIPv4ICMPChecksum(ipv4, icmp,
+				unsafe.Pointer(uintptr(unsafe.Pointer(icmp))+ICMPLen)))
 		}
 	} else if ipv6 != nil {
 		tcp, udp, _ := p.ParseAllKnownL4ForIPv6()
 		if tcp != nil {
-			p.GetTCPForIPv6().Cksum = SwapBytesUint16(CalculatePseudoHdrIPv6TCPCksum(p.GetIPv6()))
+			p.GetTCPForIPv6().Cksum = SwapBytesUint16(CalculatePseudoHdrIPv6TCPCksum(ipv6))
 		} else if udp != nil {
-			p.GetUDPForIPv6().DgramCksum = SwapBytesUint16(CalculatePseudoHdrIPv6UDPCksum(p.GetIPv6(), p.GetUDPForIPv6()))
+			p.GetUDPForIPv6().DgramCksum = SwapBytesUint16(CalculatePseudoHdrIPv6UDPCksum(ipv6, udp))
 		}
```

Comment:
If we use -hwol flag in testScenario 2  and send icmp ipv6 packet, 2nd part will try to calculate hardware cksum and will fail, maybe it makes sense to include software variant here as for ipv4?

Reply:
Done